docs/architecture/psa-thread-safety/psa-thread-safety.md

```diff
@@ -281,40 +281,36 @@ Note that a thread must hold the global mutex when it reads or changes a slot's
 
 #### Slot states
 
-For concurrency purposes, a slot can be in one of three states:
+For concurrency purposes, a slot can be in one of four states:
 
-* UNUSED: no thread is currently accessing the slot. It may be occupied by a volatile key or a cached key.
-* WRITING: a thread has exclusive access to the slot. This can only happen in specific circumstances as detailed below.
-* READING: any thread may read from the slot.
+* EMPTY: no thread is currently accessing the slot, and no information is stored in the slot.
+* FILLING: one thread is currently loading or creating material to fill the slot, this thread is responsible for the next state transition.
+* FULL: the slot contains a key, and any thread is able to use the key after registering as a reader.
+* PENDING_DELETION: the key within the slot has been destroyed or marked for destruction, but at least one thread is still registered as a reader. No thread can register to read this slot. The slot must not be wiped until the last reader de-registers, wiping the slot by calling `psa_wipe_key_slot`.
 
-A high-level view of state transitions:
+To change `slot` to state `new_state`, a function must call `psa_slot_state_transition(slot, new_state)`.
 
-* `psa_get_empty_key_slot`: UNUSED → WRITING.
-* `psa_get_and_lock_key_slot_in_memory`: UNUSED or READING → READING. This function only accepts slots in the UNUSED or READING state. A slot with the correct id but in the WRITING state is considered free.
-* `psa_unlock_key_slot`: READING → UNUSED or READING.
-* `psa_finish_key_creation`: WRITING → READING.
-* `psa_fail_key_creation`: WRITING → UNUSED.
-* `psa_wipe_key_slot`: any → UNUSED. If the slot is READING or WRITING on entry, this function must wait until the writer or all readers have finished. (By the way, the WRITING state is possible if `mbedtls_psa_crypto_free` is called while a key creation is in progress.) See [“Destruction of a key in use”](#destruction-of-a-key-in-use).
+A counter field within each slot keeps track of how many readers have registered. Library functions must call `psa_register_read` before reading the key data witin a slot, and `psa_unregister_read` after they have finished operating.
 
-The current `state->lock_count` corresponds to the difference between UNUSED and READING: a slot is in use iff its lock count is nonzero, so `lock_count == 0` corresponds to UNUSED and `lock_count != 0` corresponds to READING.
+Library functions which operate on a slot will return `PSA_ERROR_BAD_STATE` if the slot is in an inappropriate state for the function at the linearization point.
```

Comment:
As of 6ecb9ce5fc7d77b6d54881ab4d78231fb4784f98, I'm still unhappy with the document because it talks about linearization points without justifying that one exists.

I'm not even convinced that using the concept is at all useful, but maybe? Is this concept commonly known by embedded engineers? Personally I'm used to reasoning about concurrent systems that are not coherent, so the concept linearization points doesn't come up often because there often isn't one.

Reply:
https://arm-software.github.io/psa-api/crypto/1.1/overview/conventions.html#concurrent-calls states that concurrent calls which obey a list of constraints must act as if they are executed in sequence, does this not inherently mean the functions must have linearization points (if the constraints are met)?

Is there a more accessible method of reasoning we can use to prove that we satisfy this requirement? This is the first time that I have worked on concurrent software with an engineering user base instead of an academic one, so I am still learning how things are done, and appreciate the guidance :)

Reply:
This implies that a correct implementation must have linearization points. But the reasoning in the document goes the other way round: we're constructing an implementation and we'd like to know that it meets the specification. What I'm missing is an explanation (which doesn't need to be long) that the functions designed as described here have linearization points. And maybe an explanation of what that means, because I'm not sure if our audience are familiar with the term.

Reply:
Sure, I can do that! I propose a new subsection about linearizability/linearization points: including an introduction to the terms, the properties we uphold (never change X without holding the mutex), then a list containing short arguments for where the linearization points are for specific functions and why they linearize here? Most functions won't be performed with the mutex held at all times, so it is difficult to reason without doing a case-by-case analysis.

Does this sound good to you, if so I'll start work on it?

Reply:
Sure, sounds good!

Note that we don't need to _prove_ that the design is correct. We just need to explain in high-level terms how it meets the requirements.

For what it's worth, my first reactions when I read the original version of the document were:

1. Er, “linearization points”? Does that refer to linearizability in the sense of behaving as if the system was sequential, or is that a (perhaps subtly) different concept? [After: yes, I think we do have the same concept in mind.]
2. Assuming it means what I think it means, why would linearization points exist? Are we making assumptions about memory coherency? [After: they exist because of the global lock. There is no assumption about the memory: taking/releasing the global lock will take care of memory synchronization.]

Reply:
I have just pushed the new section

Reply:
Thanks for doing the analysis of linearizability.

> functions which call both psa_start_key_creation and psa_finish_key_creation. Both of these functions have effects visible to other threads, so making import_key linearizable might require holding the mutex for the entire duration of creating a key.

So per your analysis, [key creation is not linearizable](https://github.com/Mbed-TLS/mbedtls/pull/8618#issuecomment-1858264256). I hadn't thought of that. But in retrospect, it makes sense. A failed key creation temporarily holds a resource, and during that window of time one can observe a resource exhaustion which is not observable before or after, only during. (And similarly, when opaque driver support is fully production-ready, destruction will have a temporary state where one resource is freed but another one isn't yet.)

It is unavoidable that the execution of library functions is observable in terms of available resources. If nothing else, one would expect that a library function may consume some memory, so a concurrent call might fail during its execution but would succeed before or after.

This does not violate the intent of the [specification](https://arm-software.github.io/psa-api/crypto/1.1/overview/conventions.html#concurrent-calls) that states that “Concurrent calls are performed correctly, as if the calls were executed in sequence”: the intent there was that this “as if” only concerns functional correctness, not resource availability, which the specification does not cover. (For example, the specification never offers any availability guarantee.) I will raise an issue to clarify the text though.

It is desirable that key management operations do not take the global lock. In the long term, no blocking operation or expensive calculation should take place with the global lock held. This includes all driver calls. It's good for key creation to have an initial atomic step, a final atomic step, and no interference with other threads in between (or possibly other atomic step in some cases, e.g. for driver communication or storage updates).

So please document that linearizability is only true as far as functional correctness is concerned, and not as far as resource management is concerned. Other than that I'm happy with abd8977cc15a460213afc72a02aeca778c3f2bfd.